### PR TITLE
`ComponentType` interface is introduced

### PR DESCRIPTION
**Pull request by sgrishchenko**
1) `ComponentType` interface is introduced
1) Asterisk is used for Context generic parameter bound
1) `FunctionalComponent` is deprecated (POTENTIALLY BREAKING CHANGE!!!)

### CODE REVIEW

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt [hunk 1]
```diff
@@ -37,21 +37,23 @@ external interface RErrorInfo
...
     val default: RClass<P>
 }
 
-external interface RComponentClassStatics<P : RProps, S : RState, C : RContext<Any>?> {
+external interface RComponentClassStatics<in P : RProps, S : RState, C : RContext<*>?> {
     var displayName: String?
 
-    var defaultProps: P?
+    var defaultProps: @UnsafeVariance P?
+        @Deprecated(message = "Write-only property", level = DeprecationLevel.HIDDEN)
```
Comment (turansky) [inline kotlin-react/src/main/kotlin/react/ReactComponent.kt:50]: public `get` isn't problem

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt [hunk 2]
```diff
@@ -37,21 +37,23 @@ external interface RErrorInfo
...
+        get() = definedExternally
```
Comment (turansky) [inline kotlin-react/src/main/kotlin/react/ReactComponent.kt:51]: short declaration required

**Review by turansky — CHANGES_REQUESTED** [review]

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt [hunk 3]
```diff
@@ -37,21 +37,23 @@ external interface RErrorInfo
 val RErrorInfo.componentStack: Any
     get() = asDynamic().componentStack
 
-// TODO: Should extend RComponentClassStatics, but has problems with generic params
-external interface RClass<in P : RProps> : RComponentClassStatics<RProps, RState, RContext<Any>?>
+external interface RClass<in P : RProps> : RComponentClassStatics<P, RState, RContext<*>?>
```
Comment (turansky) [inline kotlin-react/src/main/kotlin/react/ReactComponent.kt:40]: Is it `ComponentClass` alias?

**Review by turansky — CHANGES_REQUESTED** [review]

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt (hunk 2 above)
Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ReactComponent.kt:51]: Do you mean just `var defaultProps: @UnsafeVariance P?` ?

**Review by turansky — APPROVED** [review]

File: kotlin-react/src/main/kotlin/react/ComponentType.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package react
+
+external interface ComponentType<in P : RProps>
+
+// TODO: Should extend RComponentClassStatics, but has problems with generic params
```
Comment (turansky) [inline kotlin-react/src/main/kotlin/react/ComponentType.kt:5]: Solved?

**Review by turansky — APPROVED** [review]

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt (hunk 1 above)
Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ReactComponent.kt]: Reverted

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt (hunk 2 above)
Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ReactComponent.kt]: Reverted

File: kotlin-react/src/main/kotlin/react/ReactComponent.kt (hunk 3 above)
Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ReactComponent.kt]: Done

File: kotlin-react/src/main/kotlin/react/ComponentType.kt (hunk 4 above)
Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ComponentType.kt:5]: Nope

Comment (sgrishchenko) [reply kotlin-react/src/main/kotlin/react/ComponentType.kt:5]: We have discussed that current `RComponentClassStatics` is incorrect, so TODO comment still actual

**Review by turansky — APPROVED** [review]

**Comment by sgrishchenko** [comment]
These changes relates to #358

**Comment by Leonya** [comment]
Released in pre.215

**Pull request merged**